File: src/lib/scala/Slate.Entities/src/main/scala/slate/entities/core/EntityMapper.scala
/**
  * <slate_header>
  * url: www.slatekit.com
  * git: www.github.com/code-helix/slatekit
  * org: www.codehelix.co
  * author: Kishore Reddy
  * copyright: 2016 CodeHelix Solutions Inc.
  * license: refer to website and/or github
  * about: A Scala utility library, tool-kit and server backend.
  * mantra: Simplicity above all else
  * </slate_header>
  */

package slate.entities.core

import slate.common.mapper.Mapper
import slate.common.query.QueryEncoder
import slate.common.{DateTime, Model, Reflector, Strings}

import scala.reflect.runtime.universe._


/**
  * Maps an entity to sql and from sql records.
  *
  * @param model
  */
class EntityMapper(model:Model) extends Mapper(model) {

  def mapToSql(item:AnyRef, update:Boolean, fullSql:Boolean = false): String = {

    if (!_model.any)
      Strings.empty
    else
      mapFields(item, update, fullSql)
  }


  private def mapFields(item:AnyRef, update:Boolean, fullSql:Boolean = false): String = {
    var dat = ""
    var sql = ""
    var cols = ""


    val len = _model.fields.size
    for( ndx <- 0 until len)
    {
      val mapping = _model.fields(ndx)
      val propName = mapping.name
      val colName = getColumnName(mapping.storedName)
      val include = !Strings.isMatch(propName, "id")

      if(include)
      {
        val data = if( mapping.dataType == _tpeString )
        {
          val sVal = Reflector.getFieldValue(item, mapping.name).asInstanceOf[String]
          val sValFinal = Strings.valueOrDefault(sVal, "")
         "'" + QueryEncoder.ensureValue(sValFinal) + "'"
        }
        else if(mapping.dataType == typeOf[Boolean])
        {
          val bVal = Reflector.getFieldValue(item, mapping.name).asInstanceOf[Boolean]
          if(bVal)  "1" else "0"
        }
        else if(mapping.dataType == typeOf[Int])
        {
          val iVal = Reflector.getFieldValue(item, mapping.name).asInstanceOf[Int]
          iVal.toString
        }
        else if(mapping.dataType == typeOf[Short])
        {
          val iVal = Reflector.getFieldValue(item, mapping.name).asInstanceOf[Short]
          iVal.toString
        }
        else if(mapping.dataType == typeOf[Long])
        {
          val lVal = Reflector.getFieldValue(item, mapping.name).asInstanceOf[Long]
          lVal.toString
        }
        else if(mapping.dataType == typeOf[Double])
        {
          val dVal = Reflector.getFieldValue(item, mapping.name).asInstanceOf[Double]
          dVal.toString
        }
        else if(mapping.dataType == typeOf[DateTime])
        {
          val dtVal = Reflector.getFieldValue(item, mapping.name).asInstanceOf[DateTime]
          "'" + dtVal.toStringMySql() + "'"
        }
        else // Object
        {
          val objVal = Reflector.getFieldValue(item, mapping.name)
          val data = Option(objVal).fold("")( obj => obj.toString())
          "'" + QueryEncoder.ensureValue(data) + "'"
        }

        // Setup the inserts
        val isLastField = ndx == _model.fields.size - 1
        if(!update)
        {
          cols += colName
          dat += data
          if(!isLastField)
          {
            cols += ","
            dat += ","
          }
        }
        else
        {
          sql += colName + "=" + data
          if(!isLastField)
          {
            sql += ","
          }
        }
      }
    }

    if(update)
    {
      sql = " " + sql
    }
    else
    {
      cols = "(" + cols + ") "
      sql = cols + "VALUES (" + dat + ")"
    }
    if(!fullSql)
      sql
    else if(update)
      s"update ${_model.name} set " + sql + ";"
    else
      s"insert into ${_model.name} " + sql + ";"
  }


  def getColumnName(name:String) : String =
  {
    "`" + name + "`"
  }
}

File: src/lib/kotlin/slatekit-entities/src/main/kotlin/slatekit/entities/mapper/EntityMapper.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */


package slatekit.entities.mapper

import slatekit.common.Record
import slatekit.common.crypto.Encryptor
import slatekit.common.data.DataAction
import slatekit.common.data.DataType
import slatekit.meta.models.Model
import slatekit.data.Mapper
import slatekit.common.data.Values
import slatekit.data.core.Meta
import slatekit.data.encoders.Encoders
import kotlin.reflect.KClass

/**
 * Maps an entity to sql and from sql records.
 *
 * @param model
 */
open class EntityMapper<TId, T>(val model: Model,
                                val meta:Meta<TId, T>,
                                val idClass: KClass<TId>,
                                val enClass: KClass<T>,
                                val settings: EntitySettings = EntitySettings(true),
                                val encoder: Encoder<TId, T> = EntityEncoder(model, meta, settings, null, Encoders()),
                                val decoder: Decoder<TId, T> = EntityDecoder(model, meta, idClass, enClass, settings, null))
    : Mapper<TId, T> where TId : kotlin.Comparable<TId>, T:Any  {

    /**
     * Gets the table column name mapped to the field name
     */
    override fun column(field:String): String = model.lookup[field]?.storedName ?: field

    /**
     * Gets the data type of the field
     */
    override fun datatype(field:String): DataType {
        return when(val mappedField = model.lookup[field]){
            null -> throw Exception("Field not mapped! model=${model.name}, field=$field")
            else -> mappedField.dataTpe
        }
    }

    /**
     * Gets the encode values for the model to be used for building a sql statement
     */
    override fun encode(model: T, action: DataAction, enc: Encryptor?): Values {
        return encoder.encode(model, action, enc)
    }

    /**
     * Decodes the record into the model
     */
    override fun decode(record: Record, enc: Encryptor?): T? {
        return decoder.decode(record, enc)
    }
}
